Feature/vegman support

nest1ing opened the pull request:
This PR brings a support for Vegman's firmware

nest1ing committed:
meson: link stdc++fs

Links the library with more appropriate way.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
Make OpenBMC implementation optional

The OpenBMC implementation is specific for obmc-phosphor-image only.
On our new hardware we have another flash layout and this implementation
is not usable.

This commit makes the OpenBMC implementation optional.
It may be chosen by `-Dbmc-image-type=obmc-phosphor-image` meson
argument.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
Move reboot-guard support to main class

While reboot-guard is enabled, it should work with any implementation.
This moves the reboot-guard support to `FwUpdate` class which is used
any case.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
Add intel-platforms support

This commit brings the support for intel-platforms firmware type.
To enable it use command:
```
meson -Dbmc-image-type=intel-platforms build
```

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

AlexanderAmelkin commented on meson.build:
```diff
@@ -72,9 +69,21 @@ conf.set_quoted('HASH_FILE_NAME', get_option('hash-file-name'))
+    conf.set_quoted('OPENBMC_WHITELIST_FILE_NAME',
+                    get_option('openbmc-whitelist-file-name'))
+    fwupdate_sources += [
+        'src/obmc-phosphor-image.cpp',
```
Have you considered calling the files uniformly like, say, `src/image-obmc.cpp` and `src/image-intel.cpp` ?

nest1ing replied:
It is an image type from the Yocto.
I can rename it if you insist.

AlexanderAmelkin replied:
I understand that it's the name of the image, that's why I don't say anything about these names in other places.
However, for the file names I think it is better if similar files are located next to each other in an alphabetically sorted file list. That's why I propose to rename them like that.

AlexanderAmelkin replied:
I mean, you are not directly using the image name from Yocto to include/link the implementation, are you? Then there is no reason to keep the name of the file identical to the name of the image.

nest1ing replied:
Ok, I will rename it.

In additions, I've found at https://github.com/openbmc/docs/blob/master/cpp-style-and-conventions.md:
> C++ files should be named with lower_snake_case.

AlexanderAmelkin commented on src/intel-platforms.cpp:
```diff
@@ -0,0 +1,112 @@
+    Tracer tracer("Clear writable partitions");
+    std::ignore = exec("flash_erase /dev/mtd/rwfs 0 0");
+    std::ignore = exec("flash_erase /dev/mtd/sofs 0 0");
+    std::ignore = exec("flash_erase /dev/mtd/u-boot-env 0 0");
```
Are you erasing the whole environment here and then only set the boot address?
What happens to other variables like `bootargs`, `baudrate`, `ethaddr`, etc. ?

nest1ing replied:
It is a factory resetting. So, the boot address makes sense only. All other fields will be dropped to their default values.

AlexanderAmelkin replied:
Who sets the default values? U-boot will not do so if the environment is non-empty.

nest1ing replied:
`ethaddr` is not set by default and should be cleaned on resetting.
`baudrate`, `bootargs` and so on are set to their default values by `u-boot`/`u-boot-tools` even if the partition is empty.

In fact, the original firmware image has empty `u-boot-env` partition.
So, this is a trying to reproduce this stage.
But the boot address should be saved, because it affects which image will be loaded.

artemsen commented on src/dbus.cpp:
```diff
@@ -50,3 +50,31 @@ void startUnit(const std::string& unitname)
+    {
+        systemBus.call_noreply(req);
+    }
+    catch (const sdbusplus::exception::SdBusError& e)
```
Unused variable `e`

nest1ing replied:
fixed

artemsen commented on src/intel-platforms.cpp:
```diff
@@ -0,0 +1,112 @@
+ */
+static void set_boot_address(size_t address)
+{
+    std::ignore = exec("fw_setenv bootcmd bootm %08x", address);
```
Do you really need std::ignore here? Looks like syntax sugar.

nest1ing replied:
Yes, it is a syntax sugar. But it suppresses the warnings about unused returning results.

AlexanderAmelkin commented on src/intel-platforms.cpp:
```diff
@@ -0,0 +1,112 @@
+    return true;
+}
+
+bool IntelPlatformsUpdater::is_file_belong(const fs::path& file) const
```
`belong` is a verb, not a noun or adjective. You can't say "file is belong", you can only say "file belongs". Also, it's not clear to me where exactly does it belong. Maybe call the function like `is_file_correct()` or `is_file_flashable()` ?

Also, it would be good to fully follow obmc coding style with all their camelCase things, and not mix CamelCase::with_snake_case.

nest1ing replied:
This was discussed in one of the previous reviews. But I can't find it in the history. Probably, it was completely deleted.

Well, I like `isFileFlashable`. I will rewrite it.

nest1ing committed:
Code style fix: use lowerCamelCase

This renames all varibales, functions and methods from lower_snake_case
to lowerCamelCase.

This also renames `is_file_belong` method to `isFileFlashable`.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
rename updaters files

This renames the files with updater implementations with a common format.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

nest1ing committed:
fix exit status when manual reboot required

When fwupdate unable to reboot the BMC is not a failure.
This commit fixes the return code for this case.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

artemsen commented on src/main.cpp:
```diff
@@ -108,7 +108,8 @@ void reboot(bool interactive)
     if (manualReboot)
     {
-        throw FwupdateError("The BMC needs to be manually rebooted.");
+        fprintf(stdout, "The BMC needs to be manually rebooted.\n");
```
`puts()`?

AlexanderAmelkin replied:
I wouldn't mix different output methods, but I agree that simple `printf` instead of `fprintf(stdout` would be better.

artemsen commented on src/main.cpp:
```diff
@@ -108,7 +108,8 @@ void reboot(bool interactive)
     {
-        throw FwupdateError("The BMC needs to be manually rebooted.");
+        fprintf(stdout, "The BMC needs to be manually rebooted.\n");
+        fflush(stdout);
```
`printf` flushes at `\n`

artemsen commented on src/image_openbmc.cpp:
```diff
@@ -64,7 +41,7 @@ void OpenBmcUpdater::do_install(const fs::path& file)
 }
 
-bool OpenBmcUpdater::do_after_install(bool reset)
+bool OBMCPhosphorImageUpdater::doAfterInstall(bool reset)
```
Since you are using CamalCase, it should be `ObmcPhoshor...`

AlexanderAmelkin replied:
According to the coding style, acronyms must be same-case and class names should be UpperCamelCase. So `OBMCPhosphor...` is the correct name for a class.

artemsen commented on src/image_openpower.cpp:
```diff
@@ -196,17 +196,17 @@ void OpenPowerUpdater::do_before_install(bool reset)
 {
     // NOTE: This process may take a lot of time and we want to show the
     //       progress from original pflash output.
     fprintf(stdout, "Writing %s ... \n", file.filename().c_str());
```
`printf` uses stdout.

artemsen commented on src/image_openpower.cpp:
```diff
@@ -196,17 +196,17 @@ void OpenPowerUpdater::do_before_install(bool reset)
     // NOTE: This process may take a lot of time and we want to show the
     //       progress from original pflash output.
     fprintf(stdout, "Writing %s ... \n", file.filename().c_str());
     fflush(stdout);
```
`\n` is already leads to flush

AlexanderAmelkin replied:
for fwupdate it could be useful to disable buffering for stdout completely using `setvbuf` at the beginning of `main()`.

nest1ing replied:
The man pages said:
> The stream stdout  is  line-buffered when  it  points  to  a terminal.  Partial lines will not appear until
> fflush(3) or exit(3) is called, or a newline  is  printed.   This  can produce  unexpected  results,  especially
> with debugging output.

I'm not sure that a new line and  the `\n` character at the end of previous line are the same thing.
Probably you are right and this code is a overreact, but it seems to me it is appropriate.

And I agree with @AlexanderAmelkin, all these calls may be replaced with a disabling the stdout buffering.

nest1ing committed:
Disable stdout buffering

This commit removes all fflush calls and disables the stdout buffering.

Signed-off-by: Alexander Filippov <a.filippov@yadro.com>

artemsen commented on src/main.cpp:
```diff
@@ -108,8 +108,7 @@ void reboot(bool interactive)
     {
-        fprintf(stdout, "The BMC needs to be manually rebooted.\n");
-        fflush(stdout);
+        printf("The BMC needs to be manually rebooted.\n");
```
Why don't you use `puts`? It works faster if don't need any formatting.

nest1ing replied:
- to avoid the mixing of several functions for the same action
- it is more familiar for me
- `puts` will not provide a great benefit
